maitake/src/wait/map.rs

```diff
@@ -0,0 +1,926 @@
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Node")
+            .field("links", &self.links)
+            .field("waker", &self.waker)
+            .finish()
+    }
+}
+
+/// The state of a [`Waiter`] node in a [`WaitMap`].
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[repr(u8)]
+enum WaitState {
+    /// The waiter has not yet been enqueued.
+    ///
+    /// The number of times [`WaitMap::wake_all`] has been called is stored
+    /// when the node is created, in order to determine whether it was woken by
+    /// a stored wakeup when enqueueing.
+    ///
+    /// When in this state, the node is **not** part of the linked list, and
+    /// can be dropped without removing it from the list.
+    Start,
+
+    /// The waiter is waiting.
+    ///
+    /// When in this state, the node **is** part of the linked list. If the
+    /// node is dropped in this state, it **must** be removed from the list
+    /// before dropping it. Failure to ensure this will result in dangling
+    /// pointers in the linked list!
+    Waiting,
+
+    /// The waiter has been woken.
+    ///
+    /// When in this state, the node is **not** part of the linked list, and
+    /// can be dropped without removing it from the list.
+    Completed,
+}
+
+/// The queue's current state.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[repr(u8)]
+enum State {
+    /// No waiters are queued, and there is no pending notification.
+    /// Waiting while the queue is in this state will enqueue the waiter
+    Empty = 0b00,
+
+    /// There are one or more waiters in the queue. Waiting while
+    /// the queue is in this state will not transition the state. Waking while
+    /// in this state will wake the appropriate waiter in the queue; if this empties
+    /// the queue, then the queue will transition to [`State::Empty`].
+    Waiting = 0b01,
+
+    // TODO(AJM): We have a state gap here. Is this okay?
+    /// The queue is closed. Waiting while in this state will return
+    /// [`Closed`] without transitioning the queue's state.
+    ///
+    /// *Note*: This *must* correspond to all state bits being set, as it's set
+    /// via a [`fetch_or`].
+    ///
+    /// [`Closed`]: crate::wait::Closed
+    /// [`fetch_or`]: core::sync::atomic::AtomicUsize::fetch_or
+    Closed = 0b11,
+}
+
+#[derive(Clone, Debug)]
+enum Wakeup {
+    /// The Waiter has been created, but no wake has occurred. This should
+    /// be the ONLY state while in `WaitState::Start`
+    Empty,
+
+    /// The Waiter has moved to the `WaitState::Waiting` state. We now
+    /// have the relevant waker, and are still waiting for data. This
+    /// corresponds to `WaitState::Waiting`.
+    Waiting(Waker),
+
+    /// The Waiter has received data, and is waiting for the woken task
+    /// to notice, and take the data by polling+completing the future.
+    /// This corresponds to `WaitState::Completed`.
+    DataReceived,
```

Comment:
a thought (which is fine to save for a follow-up): i feel like it would make sense to actually store the data in this variant, instead of having a separate `UnsafeCell<MaybeUninit<V>>`? since the `Wakeup` is essentially responsible for determining whether there's a value in there or not. i don't _think_ there's a performance/space improvement from not just storing it here?

Reply:
That's a good point! I've added a TODO comment.